sdk/search/search-documents/src/indexModels.ts

```diff
@@ -685,3 +722,136 @@ export type IndexDocumentsAction<T> = {
 } & Partial<T>;
 
 // END manually modified generated interfaces
+
+// Utility types
+
+export type UnionToIntersection<U> =
+  // Distribute members of U into parameter position of a union of functions
+  (
+    U extends unknown ? (_: U) => unknown : never
+  ) extends // Infer the intersection of the members of U as a single intersected parameter type
+  (_: infer I) => unknown
+    ? I
+    : never;
+
+// Types that should not be included in SelectFields recursion
+export type ExcludedODataTypes = Date | GeographyPoint;
+
+/**
+ * Produces a union of valid Cognitive Search OData $select paths for T
+ * using a post-order traversal of the field tree rooted at T.
+ */
+export type SelectFields<T extends object> = T extends Array<infer U>
```

Comment:
I wonder if giving more verbose names to the template variables may help readability. What do you think about:
```ts
export type SelectFields<Obj extends object> = Obj  extends Array<infer Elem>
  ? // Allow selecting fields only from elements which are objects
    NonNullable<Elem> extends object
    ? SelectFields<NonNullable<Elem>>
    : never
  : {
      // Only consider string keys
      [Key in Exclude<keyof Obj , symbol | number>]: NonNullable<Obj [Key ]> extends object
        ? NonNullable<Obj [Key ]> extends ExcludedODataTypes
          ? // Excluded, so don't recur
            Key 
          : SelectFields<NonNullable<Obj [Key]>> extends infer NextPaths
          ? // Narrow NextPaths' type for template literal
            NextPaths extends string
            ? // Union this key with all the next paths separated with '/'
              Key | `${Key}/${NextPaths}`
            : // We didn't infer any nested paths, so just use this key
              Key
          : never
        : // Not an object, so can't recur
          Key ;
    }[Exclude<keyof Obj , symbol | number>];
```

If we think the readability improvement is marginal, maybe it is not worth changing